Strings finally in Flash!

crankyoldgit opened the pull request:
Finally convince the compiler to store the text strings into flash.

Saves approx 2k of Global ram, for a trade-off of between ~0-0.5k of extra flash space used.

e.g. _(updated)_
* IRMQTTServer example code:
  - Before:
RAM:   [=====     ]  54.1% (used 44344 bytes from 81920 bytes)
Flash: [=====     ]  54.2% (used 566209 bytes from 1044464 bytes)
Bin file size = 570368
  - After:
RAM:   [=====     ]  51.3% (used 41992 bytes from 81920 bytes)
Flash: [=====     ]  54.2% (used 566201 bytes from 1044464 bytes)
Bin file size = 570352

* IRrecvDumpV2 example code:
  - Before:
RAM:   [====      ]  37.9% (used 31044 bytes from 81920 bytes)
Flash: [====      ]  35.6% (used 372025 bytes from 1044464 bytes)
Bin file size = 376176
  - After:
RAM:   [====      ]  35.5% (used 29072 bytes from 81920 bytes)
Flash: [====      ]  35.7% (used 372525 bytes from 1044464 bytes)
Bin file size = 376672

Fixes #1614
Fixes #1493

Co-authored with @mcspr

crankyoldgit committed:
Strings finally in Flash!
Finally convince the compiler to store the text strings into flash.

Saves approx 1k of Global ram, and 1-1.5k of Flash space.
Plus an overall reduction of ~1k on the resulting bin file size.

e.g.
* IRMQTTServer example code:
  - Before:
    RAM:   [=====     ]  53.2% (used 43560 bytes from 81920 bytes)
    Flash: [=====     ]  53.1% (used 554500 bytes from 1044464 bytes)
    Bin file size = 558656
  - After:
    RAM:   [=====     ]  52.0% (used 42572 bytes from 81920 bytes)
    Flash: [=====     ]  53.0% (used 553448 bytes from 1044464 bytes)
    Bin file size = 557600
* IRrecvDumpV2 example code:
  - Before:
    RAM:   [====      ]  37.0% (used 30348 bytes from 81920 bytes)
    Flash: [====      ]  35.4% (used 369644 bytes from 1044464 bytes)
    Bin file size = 373792
  - After:
    RAM:   [====      ]  35.9% (used 29372 bytes from 81920 bytes)
    Flash: [====      ]  35.2% (used 368036 bytes from 1044464 bytes)
    Bin file size = 372192

Fixes #1614
Fixes #1493

crankyoldgit commented:
FYI @Pecacheu & @mcspr

NiKiZe reviewed (approved):
This is almost a "replace" of `PROGMEM char* x` with `PROGMEM char x[]`
and then same type change of the `extern` declarations.

crankyoldgit commented:
> This is almost a "replace" of `PROGMEM char* x` with `PROGMEM char x[]` and then same type change of the `extern` declarations.

That is pretty much correct. Outside of that, it adds `#include "IRtext.h"` to `IRtext.cpp` and some notes about why it is the way it is, for future generations of frustrated Arduino ESP8266 programmers. Le Sigh!

crankyoldgit commented:
I mean seriously ... `char *array` and `char array[]` are equivilent AFAIK, but clearly not in Arduino-land.

mcspr commented:
at least one thing will break here - bulk of IRtext strings is used in the IRac::strTo* with strcasecmp, which will crash. will try to find the other calls with '\<string\>' funcs...

specific to IRac, these should be strcasecmp_P instead and the test framework updated with something like these:
https://github.com/esp8266/Arduino/blob/9d024d17fd18dffa569fa646b5326b5e8f58b0d5/tests/host/sys/pgmspace.h#L57

crankyoldgit committed:
Replace all ESP8266 `str*()`s with `_P` variants.

Make sure any strlen() & strcasecmp()'s are replaced with strlen_P() & strcasecmp_P()'s on the ESP8266 platform.

AFAIUI, the _P variants will handle non-flash located strs as well.

Add a macro structure to allow overrides and allow minimal code changes.

Note: Entire code base was checked for any `_P` possible replacements.
* Example code was excluded (and it doesn't handle the flash stored strings anyway) i.e. Its safe.
* `memcpy()`s exist but sources are not stored in flash at present.

Convert some strings to Flash based via the `F()` macro.
Ref: https://github.com/crankyoldgit/IRremoteESP8266/pull/1623#issuecomment-933633486

crankyoldgit commented:
> at least one thing will break here - bulk of IRtext strings is used in the IRac::strTo* with strcasecmp, which will crash. will try to find the other calls with '<string>' funcs...
> 
> specific to IRac, these should be strcasecmp_P instead and the test framework updated with something like these: https://github.com/esp8266/Arduino/blob/9d024d17fd18dffa569fa646b5326b5e8f58b0d5/tests/host/sys/pgmspace.h#L57

Latest commit to the branch should address that now. Let me know if it doesn't.

I couldn't find any other likely occurrences anywhere else in the code base, so fingers crossed, I think I may have them all.

crankyoldgit committed:
Oops. Change F() to PSTR().

NiKiZe commented on src/IRac.cpp:
```diff
@@ -3304,12 +3307,12 @@ int16_t IRac::strToModel(const char *str, const int16_t def) {
     return true;
   else if (!STRCASECMP(str, kOffStr) ||
-           !STRCASECMP(str, F("0")) ||
+           !STRCASECMP(str, PSTR("0")) ||
```
How large is this address? Is it maybe cheaper to store this one byte as is? (Not saying we should, just trying to see corner cases)

crankyoldgit replied:
Good point. I'll look into it
Though, technically it is two bytes. `"0"` is `{0x30, 0x00}`

So, if the alignment is a 16bit DWORD, we should be fine, if 32bit, then yeah, it is wasted space.

NiKiZe replied:
On 0 and 1 there is no need for casecmp, if that helps.

crankyoldgit replied:
Looks like DWORDs are 32bit. Thus we are wasting 2 bytes of flash.
The other option is it is stored in ram (not 100% sure on this), thus we lose 2 bytes of a much more finite resource.

crankyoldgit replied:
> On 0 and 1 there is no need for casecmp, if that helps.

Yeah, I figured that also. I'm not sure it helps.

crankyoldgit replied:
FWIW, rough calc of wasted space:
_There are approx 170+ flash-stored strings. Rounding up to 200 for simplicity/safety.
 Worst case alignment scenario is 3 bytes wasted per string. So 600 bytes wasted.
 Average case (most likely) is 2 bytes wasted, so 400 bytes._

If we packed them like the protocol names and traded off cpu & memory cycles, plus loss of code niceness/flexibility, we could lose only 4 bytes total (worst case) plus additional code overhead.

If we were saving IRAM, then yes it'd be worth it; Ram, probably; Flash, probably not.

We eat about 1-2k of flash per additional protocol, so not that worried for 400-ish bytes of flash. Yet.

crankyoldgit commented:
FYI. This utterly breaks IRrecvDumpV2 etc. So don't use this. Depending on the fix, I may go with plan B

mcspr commented:
> FYI. This utterly breaks IRrecvDumpV2 etc. So don't use this. Depending on the fix, I may go with plan B

Ok... This is embarrassing, because only part of the WString deals flashmem & ram string handling and it will fail with operator+= / concat. The first crash is at here:
```
0x40203c9b: String& String::operator+=<__FlashStringHelper const*>(__FlashStringHelper const* const&) at C:\Users\max\.platformio\packages\framework-arduinoespressif8266\cores\esp8266/WString.h:134
  \-> inlined by: resultToHumanReadableBasic(decode_results const*) at C:\Users\max\Documents\PlatformIO\Projects\IRremoteESP8266\src/IRutils.cpp:384
```
b/c of the `strlen` call trying to access flash address (see String::concat), but I was looking at the different operator and it got lost in the linked issue 🤦 

`FPSTR(kProtocolStr)` will do `reinterpret_cast<const __FlashStringHelper*>(kProtocolStr)` and things will work though. But, since this is another new code, let me think on alternative a bit more.
(and maybe patch the esp8266 Core as well)

mcspr commented:
Just as a quick example
https://github.com/mcspr/IRremoteESP8266/tree/strings_in_flash_flashstringtype
(however... strange that looks :)

But it is slightly larger, will try actually looking at the resulting binary tomorrow
```
(penv) ~\D\P\P\I\e\IRrecvDumpV2> pio run -e nodemcuv2 -t checkprogsize
...
RAM:   [====      ]  35.5% (used 29068 bytes from 81920 bytes)
Flash: [====      ]  35.7% (used 372553 bytes from 1044464 bytes)
```
(...probably b/c of some implicit string conversions...)

mcspr commented:
> (...probably b/c of some implicit string conversions...)

And one more thing

```diff
diff --git a/src/IRutils.cpp b/src/IRutils.cpp
index 1d8e4b5..0c18209 100644
--- a/src/IRutils.cpp
+++ b/src/IRutils.cpp
@@ -535,6 +535,11 @@ namespace irutils {
     return addLabeledString((value ? kOnStr : kOffStr), label, precomma);
   }

+  String addBoolToString(const bool value, const __FlashStringHelper* label,
+                         const bool precomma) {
+    return addBoolToString(value, String(label), precomma);
+  }
+
   /// Create a String with a colon separated labeled Integer suitable for
   /// Humans.
   /// e.g. "Foo: 23"
diff --git a/src/IRutils.h b/src/IRutils.h
index fdda6d7..795e25b 100644
--- a/src/IRutils.h
+++ b/src/IRutils.h
@@ -48,6 +48,8 @@ float fahrenheitToCelsius(const float deg);
 namespace irutils {
   String addBoolToString(const bool value, const String label,
                          const bool precomma = true);
+  String addBoolToString(const bool value, const __FlashStringHelper* label,
+                         const bool precomma = true);
   String addIntToString(const uint16_t value, const String label,
                         const bool precomma = true);
   String addSignedIntToString(const int16_t value, const String label,
```

Size does go down to `Flash: [====      ]  35.3% (used 368389 bytes from 1044464 bytes)`, but it feels like there's a bug somewhere in the `String` itself

crankyoldgit commented:
> b/c of the `strlen` call trying to access flash address (see String::concat), but I was looking at the different operator and it got lost in the linked issue 🤦
> 
> `FPSTR(kProtocolStr)` will do `reinterpret_cast<const __FlashStringHelper*>(kProtocolStr)` and things will work though. But, since this is another new code, let me think on alternative a bit more. (and maybe patch the esp8266 Core as well)

Yeah, I found that out already. Where I use `strlen`(well, `strlen_P`) via a pointer to an arbitary spot in a `PSTR`, it crashes the ESP8266, thus I had to replace/implement it via using the pgm_blah functions to get one byte at a time.
I've gotten it to progress a lot further by wrapping most `kBlahStr`s with `FPSTR(kBlahStr)`s or `PSTR(kBlahStr)` but this is getting ugly **_fast_**.
i.e. I'm happy to do it inside the library, but if some existing ESP8266 user has relied on accessing `kBlahStr` it's going to cause them no end of grief working out why.

If I am going to cause that much pain, I might as well hide the strings behind an access function so we are sure we are getting a correct/safe type to a user.

/me shakes fist at the ESP8266 architecture!

crankyoldgit commented:
> Just as a quick example
> https://github.com/mcspr/IRremoteESP8266/tree/strings_in_flash_flashstringtype
> (however... strange that looks :)

Okay. You've convinced me to stick with/go with that approach for now. That's less messy than my experimental code was turning into.

crankyoldgit committed:
wip (#1629)

Co-authored-by: Maxim Prokhorov <prokhorov.max@outlook.com>

crankyoldgit committed:
Fix introduced linter errors/issues.

crankyoldgit committed:
Fix code compilation errors and unit test fails.

crankyoldgit committed:
Fix doxygen warnings.

crankyoldgit committed:
Update IRtext.h generator.

mcspr commented on src/IRutils.cpp:
```diff
@@ -117,16 +140,24 @@ decode_type_t strToDecodeType(const char * const str) {
   } else {
-    for (uint16_t i = 0; i <= protocol && strlen(ptr); i++) {
-      if (i == protocol) {
-        result = ptr;
```
btw I have tried the typeToString & stringToType, after you mentioned issue with strlen_P I assumed something must be broken somewhere around here. `strToDecodeType` itself was successful, but not finding the string in the kProtocol blob caused it to go to `typeToString` and crash at this line. `result = FPSTR(ptr)` would've done it here

crankyoldgit replied:
I'm not sure/convinced as I remember testing that. IIRC from my testing, because `ptr` could be non-32bit aligned, it would cause the ESP to crash. The `kAllProtocolNamesStr` would be 32bit aligned array, but as `ptr` could point to potentially any char in the array, setting `strlen` or `strlen_P` on it made it blow up.
i.e. `strlen_P(addr_in_flash_not_32bit_aligned)` == Boom.

Hence the "hacky" char-by-char via _pgm_read_byte_ I've put in there. :-/

I had done a whole bunch of FPSTR & PSTRs to get it to work in my local git client to get things to mostly work.
e.g. I had IRrecvDumpV2 working except for some AC protocols, prior to merging in your stuff. I abandoned that work.

Anyway, I've effectively merged your branch in, and cleaned up a number of issue simple issues, and gotten it to past the "unix" tests, now to get it to compile on(for) an actual ESP chip.
i.e. platformio run just crapped the sheets.

mcspr replied:
strange :/ the strnlen_P underneath does u32 aligned reads (hence the name). before noticing the assignment issue I have checked pointer calculation and the strlen calls results, and they all worked correctly

thanks! I have missed the tests and std::string differences.

mcspr replied:
I forcibly rebased on top of the new version. 2nd const was redundant in the `reinterpret_cast<const char* const>`, and we don't actually need flashstringhelper with esp32 or tests
https://github.com/crankyoldgit/IRremoteESP8266/compare/strings_in_flash...mcspr:strings_in_flash_flashstringtype
(https://github.com/crankyoldgit/IRremoteESP8266/commit/3b9324bd8cea772a2270632f6787a87a489b1497 and https://github.com/crankyoldgit/IRremoteESP8266/commit/984f3fef8d5f90063f098d49399ad2e3b74064cf)

I also reverted this specific case of type<->string to check strlen once again with esp8266:
```cpp
#include <Arduino.h>
#include <StreamString.h>
void setup() {
  Serial.begin(115200);
}
void loop() {
  if (Serial.available()) {
    StreamString input;
    input.reserve(32);
    Serial.sendUntil(input, '\n');
    Serial.printf("> \"%s\" %d\n", input.c_str(), strToDecodeType(input.c_str()));
  }
}
```
Does all the decoding ok, no crashes so far

crankyoldgit replied:
Awesome. I will undo my hacked version then.

crankyoldgit committed:
Strings in flash flashstringtype (#1631)

* fix esp32 build flashstringhelper casts

* try to revert the original type<->string, use correct types

Authored-by: Maxim Prokhorov <prokhorov.max@outlook.com>

crankyoldgit committed:
Fix linter issues.
Fix:
```
src/IRtext.cpp:29:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/IRtext.cpp:33:  Lines should be <= 80 characters long  [whitespace/line_length] [2]

```

crankyoldgit commented:
Okay. I've tested `IRrecvDumpV2` & `IRMQTTServer` on a Nodemcu & a d1_mini. All seem to work.

crankyoldgit committed:
Minor code cleanup.

crankyoldgit commented:
FYI, this now saves ~2k of global ram. We lost a bit on the flash space, but I think I have some ideas on how to claw that back.

Updated figures:

* IRMQTTServer example code:
  - Before:
RAM:   [=====     ]  54.1% (used 44344 bytes from 81920 bytes)
Flash: [=====     ]  54.2% (used 566209 bytes from 1044464 bytes)
Bin file size = 570368
  - After:
RAM:   [=====     ]  51.3% (used 41992 bytes from 81920 bytes)
Flash: [=====     ]  54.2% (used 566201 bytes from 1044464 bytes)
Bin file size = 570352

* IRrecvDumpV2 example code:
  - Before:
RAM:   [====      ]  37.9% (used 31044 bytes from 81920 bytes)
Flash: [====      ]  35.6% (used 372025 bytes from 1044464 bytes)
Bin file size = 376176
  - After:
RAM:   [====      ]  35.5% (used 29072 bytes from 81920 bytes)
Flash: [====      ]  35.7% (used 372525 bytes from 1044464 bytes)
Bin file size = 376672

@NiKiZe Time for another re-review of the code I'm afraid. It's changed a lot.
FWIW, I'm fairly happy with it, but I'm not going to admit it's a) easy to understand & b) that I understand it fully either.

@mcspr Endless thanks mate! I'm not sure I'd have gotten this worked out in any reasonable timeframe without your help.

NiKiZe commented on src/IRac.cpp:
```diff
@@ -3167,34 +3181,34 @@ stdAc::swingv_t IRac::strToSwingV(const char *str,
-           !strcasecmp(str, D_STR_MAX D_STR_LEFT) ||       // "MaxLeft"
-           !strcasecmp(str, kMaxLeftStr))                  // "Max Left"
+  else if (!STRCASECMP(str, kLeftMaxStr) ||                     // "LeftMax"
+           !STRCASECMP(str, PSTR(D_STR_LEFT " " D_STR_MAX)) ||  // "Left Max"
+           !STRCASECMP(str, PSTR(D_STR_MAX D_STR_LEFT)) ||      // "MaxLeft"
```
Is it verified that this is working?

mcspr replied:
If the question about `#define PSTR` vs. `#define PSTR(X) X` - it works - https://godbolt.org/z/drEzfbMjq
(although, I only now noticed that actually works)

STRCASECMP works for the PSTR, since it is already `const char*` b/c it is returning pointer and [it is allowed to reinterpret_cast pointers into the same type](https://en.cppreference.com/w/cpp/language/reinterpret_cast)

*edit:* and, these do indeed work (and it's the same code as protocol parsing which I modified above)
```cpp
  assert(IRac::strToSwingH("MaxLeft") == stdAc::swingh_t::kLeftMax);
  assert(IRac::strToSwingH("Max Left") == stdAc::swingh_t::kLeftMax);
  assert(IRac::strToSwingH("Left Max") == stdAc::swingh_t::kLeftMax);
  assert(IRac::strToSwingH("LeftMax") == stdAc::swingh_t::kLeftMax);
```

crankyoldgit replied:
> Is it verified that this is working?

I didn't test them "on chip" but, as @mscpr also indicated, there is not reason they shouldn't.
Either way, they will be going away soon, as I'm going to move them to IRtext for completeness and for some other code cleanup/reductions I've got planned in the next few days.

NiKiZe commented on src/IRac.cpp:
```diff
@@ -3167,34 +3181,34 @@ stdAc::swingv_t IRac::strToSwingV(const char *str,
-           !strcasecmp(str, D_STR_MAX D_STR_RIGHT) ||      // "MaxRight"
-           !strcasecmp(str, kMaxRightStr))                 // "Max Right"
+  else if (!STRCASECMP(str, kRightMaxStr) ||                     // "RightMax"
+           !STRCASECMP(str, PSTR(D_STR_RIGHT " " D_STR_MAX)) ||  // "Right Max"
+           !STRCASECMP(str, PSTR(D_STR_MAX D_STR_RIGHT)) ||      // "MaxRight"
```
These as above?

NiKiZe reviewed (approved):
I can not find anything against this, but not tested in any way.

is there anything that we need to update in .py grabber scripts or similar?

crankyoldgit commented:
> I can not find anything against this, but not tested in any way.
> 
> is there anything that we need to update in .py grabber scripts or similar?

I updated the IRtext.h builder script. I can't immediately think of any of the python scripts that need adjusting.

Possibly the `keywords.txt` generator (which probably needs a complete overhaul or removal anyway)

crankyoldgit committed:
Update tools/mkkeywords
* Handle new constant structure for strings.
* Fix some other niggling issues.

mcspr commented:
btw, also re.
> /me shakes fist at the ESP8266 architecture!

another issue with not using ..._P aligned-access helpers is the [the alternative is pretty slow - we *could* setup a <s>hardware</s> software trap for the exception, and redirect the read to the correct value](https://github.com/esp8266/Arduino/pull/6978#issuecomment-571087360). but, it does not know the thing / pattern to access the thing we want, so the result is hit and miss most of the time.
[esp8266 Core has un-documented handler for 'generic' flash & iram access](https://github.com/esp8266/Arduino/blame/master/cores/esp8266/core_esp8266_main.cpp#L373-L375), since Core 3.0.0+ (https://github.com/esp8266/Arduino/pull/7060)
(something like this is happening with esp32 builds, [but b/c of a slightly different memory composition placing strings into data rom vs. the instruction one (ref. 1.1)](https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf#sysmem). and sw variant will possibly [will be enabled with esp8266 rtos ports]())

*edit: typo, clarify esp32 difference*

crankyoldgit commented:
FYI, the changes mentioned above have now been included in the new [v2.8.0 release](https://github.com/crankyoldgit/IRremoteESP8266/releases/tag/v2.8.0) of the library.
